src/neuroglancer/datasource/computed/backend.ts

```diff
@@ -0,0 +1,617 @@
+    const originDataType = inputSpec.dataType;
+
+    // Signal that we're now taking up memory. This value will be overwritten
+    // post-computation by a call to decodeRawChunk.
+    this.systemMemoryBytes = bufferLength * DATA_TYPE_BYTES[originDataType];
+    this.inputBuffer_ = new ArrayBuffer(this.systemMemoryBytes);
+
+    this.setupSourceChunks_();
+    this.initialized_ = true;
+  }
+
+  /**
+   * Returns a Promise that will resolve when computation completes, or reject
+   * if computation fails or is cancelled.
+   */
+  getPromise() {
+    return new Promise<void>((resolve, reject) => {
+      if (this.success_) {
+        resolve();
+        return;
+      }
+      if (this.failure_ !== undefined) {
+        reject(this.failure_);
+        return;
+      }
+      this.resolve_ = resolve;
+      this.reject_ = reject;
+    });
+  }
+
+  /**
+   * Listens to state changes on origin Chunks.
+   * @param chunk an origin Chunk.
+   */
+  stateChanged(chunk: Chunk) {
+    const volumeChunk = <VolumeChunk>chunk;
+    switch (volumeChunk.state) {
+      case ChunkState.SYSTEM_MEMORY_WORKER: {
+        this.copyOriginChunk_(volumeChunk);
+        break;
+      }
+      case ChunkState.FAILED:
+      case ChunkState.EXPIRED: {
+        this.fail_(volumeChunk.state);
+        break;
+      }
+      case ChunkState.SYSTEM_MEMORY:
+      case ChunkState.GPU_MEMORY: {
+        // The data was moved to the frontend before we could intercept it, so
+        // request it to be sent back.
+        const gridKey = gridPositionKey(volumeChunk.chunkGridPosition);
+        const chunkSize = volumeChunk.chunkDataSize!;
+        const originSource = this.source.originSource;
+        const chunkCorner = vec3.multiply(
+            vec3.create(), volumeChunk.chunkGridPosition, originSource.spec.chunkDataSize);
+
+        this.source.requestChunkData(volumeChunk, this.key!, (source: TypedArray, error: any) => {
+          if (error) {
+            console.log(this.key!, 'unable to retrieve frontend data for', volumeChunk.key!);
+            this.fail_(error);
+            return;
+          }
+          const originGridPosition = this.originGridPositions_.get(gridKey)!;
+          const originChunk = <VolumeChunk>originSource.getChunk(originGridPosition);
+          originChunk.unregisterListener(this);
+          this.originGridPositions_.delete(gridKey);
+
+          const inputSpec = this.computationParams_!.inputSpec;
+          const destination = getArrayView(this.inputBuffer_!, inputSpec.dataType);
+          const numChannels = originSource.spec.numChannels;
+          const rawSource = this.maybeDecodeBuffer_(
+              source, inputSpec.dataType, originChunk.chunkDataSize!, numChannels);
+          copyBufferOverlap(
+              chunkCorner, chunkSize, rawSource, this.inputLower_!, inputSpec.size, destination,
+              inputSpec.dataType);
+          setTimeout(() => this.checkDone_(), 0);
```

Comment:
What's the reason for the setTimeout?

Reply:
checkDone calls the computation when the buffer is full, which can take an arbitrary amount of time. The setTimeout call is to prevent it blocking the queue handling.